### PR TITLE
WIP: transcription eval wrapper

### PR DESCRIPTION
**Pull request by justinsalamon**

### CODE REVIEW

File: jams/eval.py [hunk 1]
```diff
@@ -475,3 +476,47 @@ def pattern(ref, est, **kwargs):
...
+    >>> # Load in the JAMS objects
+    >>> ref_jam = jams.load('reference.jams')
+    >>> est_jam = jams.load('estimated.jams')
+    >>> # Select the first relevant annotations
+    >>> ref_ann = ref_jam.search(namespace='pitch_hz')[0]
+    >>> est_ann = est_jam.search(namespace='pitch_hz')[0]
+    >>> scores = jams.eval.transcription(ref_ann, est_ann)
+    '''
+
+    namespace = 'pitch_hz'
```
Comment (bmcfee) [inline jams/eval.py]: Seems like you'd also want to support `pitch_midi` here, right?

It'd be ok to punt that until implementing #28 , but this particular case is pretty easy to handle IMO.

Comment (justinsalamon) [inline jams/eval.py]: Definitely. In fact, I need this functionality asap, so will be implementing it now(ish). Do you think the best way to go about it is to write the ``pitch_hz``<->``pitch_midi`` converter, and then have ``transcription()`` in ``eval.py`` detect ``pitch_midi`` annotations and convert them to ``pitch_hz`` automatically?

Comment (justinsalamon) [inline jams/eval.py]: Small note: travis will fail until we merge the ``transcription`` branch of ``mir_eval`` back into ``master`` (because right now it thinks there's no transcription module).

**Comment by justinsalamon** [comment]
I'm experiencing some odd behavior:
```python
In [1]: import jams
In [2]: refjam = jams.load('/Users/justin/dev/jams/jams/tests/fixtures/transcription_ref.jams')
In [3]: estjam = jams.load('/Users/justin/dev/jams/jams/tests/fixtures/transcription_est.jams')
In [4]: refann = refjam.search(namespace='pitch_hz')[0]
In [5]: estann = estjam.search(namespace='pitch_hz')[0]
In [6]: scores = jams.eval.transcription(refann, estann)
In [7]: scores
Out[7]: 
OrderedDict([('Precision', 0.6),
             ('Recall', 0.75),
             ('F-measure', 0.6666666666666665),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
```

Note how the metrics are the same both with offsets and without. The `transcription` branch of `mir_eval` works fine though:
```python
In [1]: import mir_eval
In [2]: refint, refpitch = mir_eval.io.load_valued_intervals('/Users/justin/Downloads/transcription/ref00.txt')
In [3]: estint, estpitch = mir_eval.io.load_valued_intervals('/Users/justin/Downloads/transcription/est00.txt')
In [4]: scores = mir_eval.transcription.evaluate(refint, refpitch, estint, estpitch)
In [5]: scores
Out[5]: 
OrderedDict([('Precision', 0.4),
             ('Recall', 0.5),
             ('F-measure', 0.4444444444444445),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
```

**Comment by bmcfee** [comment]
I'm confused -- why did commits from #105 end up in this branch?  Did you cherry-pick them over instead of rebasing off master?

**Comment by bmcfee** [comment]
> Note how the metrics are the same both with offsets and without. The transcription branch of mir_eval works fine though:

Are you sure the same data is going into both versions?  What if you construct jams objects from the CSV files and then call `eval.transcription()`?

**Comment by justinsalamon** [comment]
> I'm confused -- why did commits from #105 end up in this branch? Did you cherry-pick them over instead of rebasing off master?

I just rebased from master. Earlier I'd switched to the ``namespace_conversion`` branch a couple of times to look at some code, could that have done it?

**Comment by bmcfee** [comment]
> I just rebased from master. Earlier I'd switched to the namespace_conversion branch a couple of times to look at some code, could that have done it?

it shouldn't have.  Would you mind removing these commits from this branch and rebasing fresh from master?  Or I can do it if you prefer.

**Comment by justinsalamon** [comment]
go ahead

**Comment by bmcfee** [comment]
> go ahead

ok, done. Please pull.

**Comment by justinsalamon** [comment]
@bmcfee care to chew on this one?
```python
In [1]: import mir_eval, jams
In [2]: refint, refpitch = mir_eval.io.load_valued_intervals('/Users/justin/Downloads/transcription/ref00.txt')
In [3]: estint, estpitch = mir_eval.io.load_valued_intervals('/Users/justin/Downloads/transcription/est00.txt')
In [4]: ref_jam, ref_ann = jams.util.import_lab('pitch_hz', '/Users/justin/Downloads/transcription/ref00.txt')
In [5]: est_jam, est_ann = jams.util.import_lab('pitch_hz', '/Users/justin/Downloads/transcription/est00.txt')
In [6]: scores_mireval = mir_eval.transcription.evaluate(refint, refpitch, estint, estpitch)
In [7]: scores_jams = jams.eval.transcription(ref_ann, est_ann)
In [8]: scores_mireval
Out[8]: 
OrderedDict([('Precision', 0.4),
             ('Recall', 0.5),
             ('F-measure', 0.4444444444444445),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
In [9]: scores_jams
Out[9]: 
OrderedDict([('Precision', 0.6),
             ('Recall', 0.75),
             ('F-measure', 0.6666666666666665),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
```

**Comment by bmcfee** [comment]
> @bmcfee care to chew on this one?

1. if you do `ref_ann.data.to_interval_values()` does it look like `refint` and `refpitch`? (Same for est)
2. I made a comment over in mir_eval, but it looks like there may be shenanigans in how you're handling kwargs.  I recommend putting some debug prints into `transcription.evaluate()` over there to make sure that your parameters actually are set as you expect them to be.

**Comment by justinsalamon** [comment]
> if you do ref_ann.data.to_interval_values() does it look like refint and refpitch? (Same for est)

yes
> I made a comment over in mir_eval, but it looks like there may be shenanigans in how you're handling kwargs. I recommend putting some debug prints into transcription.evaluate() over there to make sure that your parameters actually are set as you expect them to be.

I added `print(kwargs)` just before computing the scores on both occasions (with offset, without offset) to `transcription.evaluate`, but it looks identical:
```python
In [6]: scores_mireval = mir_eval.transcription.evaluate(refint, refpitch, estint, estpitch)
{}
{'offset_ratio': None}

In [7]: scores_jams = jams.eval.transcription(ref_ann, est_ann)
{}
{'offset_ratio': None}
```

**Comment by bmcfee** [comment]
> I added print(kwargs) just before computing the scores on both occasions (with offset, without offset) to transcription.evaluate, but it looks identical:

Boy howdy, that's crazy.

Debug-print the intervals and values inside jams.eval.transcription?

**Comment by justinsalamon** [comment]
> Debug-print the intervals and values inside jams.eval.transcription?

The intervals and values look fine:
```python
In [6]: scores_mireval = mir_eval.transcription.evaluate(refint, refpitch, estint, estpitch)
{}
{'offset_ratio': None}

In [7]: scores_jams = jams.eval.transcription(ref_ann, est_ann)
[[ 0.1   0.3 ]
 [ 0.3   0.4 ]
 [ 0.5   0.6 ]
 [ 0.55  0.65]]
[220.0, 246.94200000000001, 277.18299999999999, 293.66500000000002]
[[ 0.12  0.29]
 [ 0.3   0.35]
 [ 0.5   0.6 ]
 [ 0.55  0.6 ]
 [ 0.56  0.65]]
[225.0, 246.94200000000001, 500.0, 293.66500000000002, 293.66500000000002]
{}
{'offset_ratio': None}
```
But the scores still be crazy:
```python
In [8]: scores_mireval
Out[8]: 
OrderedDict([('Precision', 0.4),
             ('Recall', 0.5),
             ('F-measure', 0.4444444444444445),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])

In [9]: scores_jams
Out[9]: 
OrderedDict([('Precision', 0.6),
             ('Recall', 0.75),
             ('F-measure', 0.6666666666666665),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
```

**Comment by justinsalamon** [comment]
Also, I made the changes you suggested on the mir_eval thread, still a no go:
```python 
In [6]: scores_mireval = mir_eval.transcription.evaluate(refint, refpitch, estint, estpitch)
{'offset_ratio': 0.2}
{'offset_ratio': None}

In [7]: scores_jams = jams.eval.transcription(ref_ann, est_ann)
[[ 0.1   0.3 ]
 [ 0.3   0.4 ]
 [ 0.5   0.6 ]
 [ 0.55  0.65]]
[220.0, 246.94200000000001, 277.18299999999999, 293.66500000000002]
[[ 0.12  0.29]
 [ 0.3   0.35]
 [ 0.5   0.6 ]
 [ 0.55  0.6 ]
 [ 0.56  0.65]]
[225.0, 246.94200000000001, 500.0, 293.66500000000002, 293.66500000000002]
{'offset_ratio': 0.2}
{'offset_ratio': None}

In [8]: scores_mireval
Out[8]: 
OrderedDict([('Precision', 0.4),
             ('Recall', 0.5),
             ('F-measure', 0.4444444444444445),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])

In [9]: scores_jams
Out[9]: 
OrderedDict([('Precision', 0.6),
             ('Recall', 0.75),
             ('F-measure', 0.6666666666666665),
             ('Precision_no_offset', 0.6),
             ('Recall_no_offset', 0.75),
             ('F-measure_no_offset', 0.6666666666666665)])
```

**Comment by justinsalamon** [comment]
Right, so with help from @bmcfee we've managed to figure out the problem (which of course wasn't actually a problem): the test case I'd hard coded included a corner case: the offset distance between ref_note2 and est_note2 is exactly 50 ms. So actually JAMS was returning the right results, since even with offsets these two notes should match, since we're using strict=False. However, mir_eval probably uses float32 rather than float64 which is used in JAMS, consequently leading to the two notes NOT matching, which I though was the correct answer, but I was wrong.

The simple "solution" is to change the test case (specifically the offset of the second note from 0.35 to 0.34): this way the test data should always give different results depending on whether we include offsets or not.

On a related note, it's probably be a good idea to add rounding to a target precision (in decimals, via np.around) in match_notes to ensure this problem doesn't resurface.

**Comment by justinsalamon** [comment]
Hey @bmcfee, travis builds are failing because it can't find the transcription module, which has now been merged into the master branch of `mir_eval` but isn't part of `mir_eval v0.2`. 

Is it possible to configure travis to install `mir_eval` directly from the master branch for testing? 

Or just generally is there an alternative way of sorting this out?

**Comment by bmcfee** [comment]
> Is it possible to configure travis to install mir_eval directly from the master branch for testing?

Yes; you just have to change the .travis_dependencies.yml script to install from github instead of pypi.

However, unless it's totally urgent, I'd prefer to not go mucking about with that if we can wait for mir_eval 0.3.0.  The reason for this is that we'll have to remember to remove that check and test again after the mir_eval release.

If this is a continuing problem, we could add mir_eval as a subtree and always benchmark against the latest master automagically.. but that seems dangerous.

**Comment by justinsalamon** [comment]
>The reason for this is that we'll have to remember to remove that check and test again after the mir_eval release.

Yeah that's what I thought.

>However, unless it's totally urgent, I'd prefer to not go mucking about with that if we can wait for mir_eval 0.3.0.

Well, it *is* urgent in that I need to use this wrapper asap, but I guess I can just use it on this branch for now. Care to CR this (it's about 8 lines) just to that once mir_eval 0.3.0 hits the market we can just merge it?

**Comment by bmcfee** [comment]
> Well, it is urgent in that I need to use this wrapper asap, but I guess I can just use it on this branch for now.

Yeah, that's what I'd recommend.  It keeps the test-merge process way simpler here.


> Care to CR this (it's about 8 lines) just so that once mir_eval 0.3.0 hits the market we can just merge it?

LGTM.  I'll thumbs-up it once mir_eval hits and we can merge.

**Comment by justinsalamon** [comment]
ok sounds good

**Comment by justinsalamon** [comment]
@bmcfee now that mir_eval is at 0.3, this PR is passing (yay). Star + Merge?

File: jams/eval.py [hunk 2]
```diff
@@ -479,3 +480,50 @@ def pattern(ref, est, **kwargs):
...
+    ref = coerce_annotation(ref, namespace)
+    est = coerce_annotation(est, namespace)
+    ref_intervals, ref_pitches = ref.data.to_interval_values()
+    est_intervals, est_pitches = est.data.to_interval_values()
+
+    ref_pitches = np.array(ref_pitches)
```
Comment (bmcfee) [inline jams/eval.py]: Why is it necessary to call `array` on this?  `to_interval_values()` already returns numpy arrays.

**Comment by bmcfee** [comment]
I made one more comment on the code, but it's minor.

Otherwise, can you rebase off master now that 0.3 support is merged?

File: jams/eval.py (hunk 2 above)
Comment (justinsalamon) [inline jams/eval.py]: `to_interval_values` returns the intervals as a numpy array, but the values (pitches in this case) as a list: https://github.com/marl/jams/blob/master/jams/core.py#L707. `mir_eval` expects both as numpy arrays, hence the conversion.

**Comment by justinsalamon** [comment]
Rebased, tests passing. Think we're good.

File: jams/eval.py (hunk 2 above)
Comment (bmcfee) [inline jams/eval.py]: right, sorry, i misread.  That makes sense.
